internal/ethereum/confirmation_reconciler.go

```diff
@@ -161,96 +329,90 @@ func (bl *blockListener) buildConfirmationQueueUsingInMemoryPartialChain(ctx con
+		if previousBlock != nil {
+			if existingConfirmation.BlockNumber.Uint64() != previousBlockNumber+1 {
+				return i18n.NewError(ctx, msgs.MsgFailedToBuildConfirmationQueue)
+			}
+			if !previousBlock.IsParentOf(existingConfirmation) {
+				return i18n.NewError(ctx, msgs.MsgFailedToBuildConfirmationQueue)
 			}
 		}
+		previousBlock = existingConfirmation
+		previousBlockNumber = existingConfirmation.BlockNumber.Uint64()
 
-		if fetchedFromExistingQueue {
-			continue
-		}
+	}
+	return nil
+}
 
-		// Fetch block from blockchain if not found in existing confirmations
-		freshBlockInfo, _, err := bl.getBlockInfoByNumber(ctx, i, false, "", "")
-		if err != nil {
-			return nil, hasNewFork, err
-		}
-		if freshBlockInfo == nil {
-			return nil, hasNewFork, i18n.NewError(ctx, msgs.MsgBlockNotAvailable)
-		}
+func (bl *blockListener) handleSpecialCases(ctx context.Context, existingConfirmations []*ffcapi.MinimalBlockInfo, txBlockInfo *ffcapi.MinimalBlockInfo, targetConfirmationCount uint64, reconcileResult *ffcapi.ConfirmationUpdateResult) (*ffcapi.ConfirmationUpdateResult, error) {
+	if targetConfirmationCount == 0 {
+		// if the target confirmation count is 0, we can immediately return a confirmed result
+		reconcileResult.Confirmed = true
+		reconcileResult.Confirmations = []*ffcapi.MinimalBlockInfo{txBlockInfo}
+		return reconcileResult, nil
+	}
 
-		fetchedBlock := &ffcapi.MinimalBlockInfo{
-			BlockNumber: fftypes.FFuint64(freshBlockInfo.Number.BigInt().Uint64()),
-			BlockHash:   freshBlockInfo.Hash.String(),
-			ParentHash:  freshBlockInfo.ParentHash.String(),
-		}
+	if uint64(len(existingConfirmations)) > targetConfirmationCount {
+		if existingConfirmations[0].Equal(txBlockInfo) {
+			// start of the existing confirmations  aligns with our latest view of the chain
 
-		// Validate parent-child relationship
-		if lastValidatedBlock != nil && !fetchedBlock.IsParentOf(lastValidatedBlock) {
-			return nil, hasNewFork, i18n.NewError(ctx, msgs.MsgFailedToBuildConfirmationQueue)
-		}
+			// attempt to validate the rest of the existing confirmations.  Starting with the target confirmation block and working forwards
 
-		// Prepend fetched block to confirmation queue
-		newConfirmationsWithoutTxBlock = append([]*ffcapi.MinimalBlockInfo{fetchedBlock}, newConfirmationsWithoutTxBlock...)
-		lastValidatedBlock = fetchedBlock
-	}
+			attemptIndex := targetConfirmationCount
 
-	// Final validation: ensure the confirmation chain connects to the transaction block
-	if len(newConfirmationsWithoutTxBlock) > 0 && !txBlockInfo.IsParentOf(newConfirmationsWithoutTxBlock[0]) {
-		return nil, hasNewFork, i18n.NewError(ctx, msgs.MsgFailedToBuildConfirmationQueue)
-	}
+			for attemptIndex < uint64(len(existingConfirmations))-1 {
+				existingConfirmationToValidate := existingConfirmations[attemptIndex]
+				blockInfo := bl.getBlockByNumberFromInMemoryPartialChain(ctx, existingConfirmationToValidate.BlockNumber.Uint64())
```

Comment:
The lock of the in-memory partial chain is embedded in the function, which means a single reconciliation logic now requires more than 1 time of lock access.